src/installer/tests/TestUtils/TestArtifact.cs

```diff
@@ -70,22 +78,23 @@ public virtual void Dispose()
 
         protected static string GetNewTestArtifactPath(string artifactName)
         {
-            var rand = new Random();
-
-            string path;
             while (true)
             {
-                path = Path.Combine(TestArtifactsPath, Path.GetRandomFileName(), artifactName);
-                Directory.CreateDirectory(path);
-                var lockPath = Path.Combine(path, ".lock");
+                var parentPath = Path.Combine(TestArtifactsPath, Path.GetRandomFileName());
+                // Create a lock file next to the target folder
+                var lockPath = parentPath + ".lock";
+                var artifactPath = Path.Combine(parentPath, artifactName);
                 try
                 {
                     File.Open(lockPath, FileMode.CreateNew, FileAccess.Write).Dispose();
-                    return path;
-                } catch
+                }
+                catch
                 {
                     // Lock file cannot be created, potential collision
```

Comment:
There should probably be some retry count and a escape/failure when nothing works. For example if the test doesn't have access right to write the file (for some weird reason) it would hang. It would better to get a clear error message (with the last exception in it). By now I expect the unexpected from CI 😉

Reply:
I usually find hangs slightly easier to diagnose -- Roslyn used to find all the test exes that were still running and create minidumps for them. But if that infrastructure doesn't work so well in runtime I can produce a log message instead.